docs/src/testing/testing_contract_internals.md

```diff
@@ -0,0 +1,187 @@
+    assert(new_block_number == old_block_number, 'Block num did not change back');
+}
+```
+### 2. Spying for events
+You can use both `starknet::emit_event_syscall` and `ContractState.emit` and the spies will capture the events,
+emitted in a `#[test]` function, if you pass the `test_address()` as a spy parameter (or spy on all events).
+
+[//]: # (TODO #851: Provide better examples)
+```rust
+use array::ArrayTrait;
+use result::ResultTrait;
+use starknet::SyscallResultTrait;
+use starknet::ContractAddress;
+use snforge_std::{ declare, ContractClassTrait, spy_events, EventSpy, EventFetcher,
+    event_name_hash, EventAssertions, Event, SpyOn, test_address };
+
+#[test]
+fn test_expect_events_simple() {
+    let test_address = test_address();
+    let mut spy = spy_events(SpyOn::One(test_address));
+    assert(spy._id == 0, 'Id should be 0');
+
+    starknet::emit_event_syscall(array![1234].span(), array![2345].span()).unwrap_syscall();
+
+    spy.assert_emitted(@array![
+        (
+            contract_address,
+            Event { keys: array![1234], data: array![2345] }
+        )
+    ]);
+
+    assert(spy.events.len() == 0, 'There should be no events left');
+}
+
+```
+
+## Testing library calls directly in the tests
+
+You can avoid deploying a mock contract, to test a `library_call` with a `LibraryCallDispatcher`.
+
+### 1. Write the contract class
+
+```rust
+ #[starknet::contract]
+mod LibraryContract {
+    use result::ResultTrait;
+    use starknet::ClassHash;
+    use starknet::library_call_syscall;
+
+    #[storage]
+    struct Storage {
+        value: felt252
+    }
+
+    #[external(v0)]
+    fn get_value(
+        self: @ContractState,
+    ) -> felt252 {
+       self.value.read()
+    }
+
+    #[external(v0)]
+    fn set_value(
+        ref self: ContractState,
+        number: felt252
+    ) {
+       self.value.write(number);
+    }
+}
+```
+### 2. Write the test
+```rust
+use result::ResultTrait;
+use starknet::{ ClassHash, library_call_syscall, ContractAddress };
+use snforge_std::{ declare };
+
+#[starknet::interface]
+trait ILibraryContract<TContractState> {
+    fn get_value(
+        self: @TContractState,
+    ) -> felt252;
+
+    fn set_value(
+        ref self: TContractState,
+        number: felt252
+    );
+}
+
+#[test]
+fn test_library_calls() {
+    let class_hash = declare('LibraryContract').class_hash;
+    let lib_dispatcher = ILibraryContractSafeLibraryDispatcher { class_hash };
+    let value = lib_dispatcher.get_value().unwrap();
+    assert(value == 0, 'Incorrect state');
+    lib_dispatcher.set_value(10);
+    let value = lib_dispatcher.get_value().unwrap();
+    assert(value == 10, 'Incorrect state');
+}
+```
+> ⚠️ **Warning**
+> 
+> This library call will write to the `test_address()` memory segment, so it can potentially **overwrite** the changes
+> you make to the memory through `contract_state_for_testing()` object and vice-versa.
```

Comment:
```suggestion
> This library call will write to the `test_address` memory segment, so it can potentially **overwrite** the changes
> you make to the memory through `contract_state_for_testing` object and vice-versa.
```